src/main/java/org/elasticsearch/indices/recovery/RecoverySource.java

```diff
@@ -356,10 +356,11 @@ public void onFailure(String source, Throwable t) {
                     }
                 });
                 try {
+                    // note, we can wait without timeout here since we added the task to the cluster service with timeout
                     latch.await();
                     if (mappingCheckException.get() != null) {
-                        logger.warn("error during mapping check, failing recovery", mappingCheckException.get());
-                        throw new ElasticsearchException("error during mapping check", mappingCheckException.get());
+                        logger.warn("{} error during mapping check, failing recovery", request.shardId(), mappingCheckException.get());
```

Comment:
+1 on adding the shardId. I  believe the exception needs to be first in the parameter list.